CMakeLists.txt

```diff
@@ -47,6 +47,12 @@ if(CMAKE_SOURCE_DIR STREQUAL CMAKE_CURRENT_SOURCE_DIR)
   project(stablehlo LANGUAGES CXX C)
   set(CMAKE_C_STANDARD 11)
   set(CMAKE_CXX_STANDARD 17)
+  
+  # Issue with LLD and -Werror. Similar to golang/go#41527, iree-org/iree#7450
+  set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wall -Werror") 
+  if (CMAKE_CXX_COMPILER_ID STREQUAL "Clang")
+    set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wno-unused-command-line-argument") 
```

Comment:
This sounds weird to me: why is -fuse-ld part of CMAKE_CXX_FLAGS instead of CMAKE_EXE_LINKER_FLAGS ?

Reply:
I have edited the PR description to include the build command. The `-fuse-ld` part comes from `-DLLVM_ENABLE_LLD=ON`, the `CMAKE_CXX_FLAGS` are separate, in the CMakeList in this case. 

At the bottom of the PR description it shows the eventual `clang++` call, editing out the files for brevity:

```
/usr/bin/clang++ -DCXX_SUPPORTS_CUSTOM_LINKER  -Wall -Werror  -fuse-ld=lld -std=gnu++17 -MD -MT ....
clang: error: argument unused during compilation: '-fuse-ld=lld' [-Werror,-Wunused-command-line-argument]
```

Perhaps there is a different configuration we should be using?

Reply:
We should fix the cmake logic upstream. The intend when I implemented it was to just use it for the linker phase: https://reviews.llvm.org/D28738
But the test was a no-op and got "fixed" later: https://reviews.llvm.org/D39274

Reply:
I roughly understand what you're suggesting, but not sure how to best implement it. I used the following, which seems to have the desired behavior (without knowing the nuances, i.e. not sure why we test before append in current setup):

```
if( LLVM_USE_LINKER )
  set(OLD_CMAKE_EXE_LINKER_FLAGS ${CMAKE_EXE_LINKER_FLAGS})
  append("-fuse-ld=${LLVM_USE_LINKER}" CMAKE_EXE_LINKER_FLAGS)
  check_cxx_source_compiles("int main() { return 0; }" CXX_SUPPORTS_CUSTOM_LINKER)
  if ( NOT CXX_SUPPORTS_CUSTOM_LINKER )
    message(FATAL_ERROR "Host compiler does not support '-fuse-ld=${LLVM_USE_LINKER}'")
  endif()
  set(CMAKE_EXE_LINKER_FLAGS ${OLD_CMAKE_EXE_LINKER_FLAGS})
  append("-fuse-ld=${LLVM_USE_LINKER}"
    CMAKE_EXE_LINKER_FLAGS CMAKE_MODULE_LINKER_FLAGS CMAKE_SHARED_LINKER_FLAGS)
endif()
```

And it seems like these problems are resolved, and unsupported linkers still error:

```
-- Performing Test CXX_SUPPORTS_CUSTOM_LINKER - Failed
CMake Error at cmake/modules/HandleLLVMOptions.cmake:309 (message):
  Host compiler does not support '-fuse-ld=notalinker'
Call Stack (most recent call first):
  CMakeLists.txt:870 (include)
```

Let me know your thoughts / any tests you think I should try. Happy to take this through the review process, should learn how to contribute to llvm-project sooner or later..

Reply:
Can we even simplify this further?
```
if( LLVM_USE_LINKER )
  append("-fuse-ld=${LLVM_USE_LINKER}"
    CMAKE_EXE_LINKER_FLAGS CMAKE_MODULE_LINKER_FLAGS CMAKE_SHARED_LINKER_FLAGS)
  check_cxx_source_compiles("int main() { return 0; }" CXX_SUPPORTS_CUSTOM_LINKER)
  if ( NOT CXX_SUPPORTS_CUSTOM_LINKER )
    message(FATAL_ERROR "Host compiler does not support '-fuse-ld=${LLVM_USE_LINKER}'")
  endif()
endif()
```

OLD_CMAKE_EXE_LINKER_FLAGS is only useful if we make it not a FATAL error right?

Reply:
But otherwise, LG :)

Reply:
Changed this PR a bit. I found the following config flags when editing the cmake file: `LLVM_ENABLE_WARNINGS` and `LLVM_ENABLE_WERROR`. When setting these flags in our cmake file, we get the desired behavior without any issues regarding LLD. Tested that StableHLO build fails with compiler warnings when these flags are set.

There are also lines that have comments regarding `-Wall`, so we likely should use these LLVM flags instead of directly adding the `-W...` compiler flags, as they seem to be safer: https://github.com/llvm/llvm-project/blob/cbf81558ab541802d1b2a859d72680173cafcb10/llvm/cmake/modules/HandleLLVMOptions.cmake#L692

I will still make the suggested change to upstream, but I don't think merging this PR is blocked by that fix.

Reply:
Created: https://reviews.llvm.org/D134206